### PR TITLE
move @CordaSerializable into it's own module to allow use without pulling in dependencies

### PR DESCRIPTION
**Pull request by roastario**

### CODE REVIEW

File: core/src/test/kotlin/net/corda/core/utilities/KotlinUtilsTest.kt [hunk 1]
```diff
@@ -73,12 +73,12 @@ class KotlinUtilsTest {
         }
     }
 
-    @CordaSerializable
+    @net.corda.annotations.serialization.CordaSerializable
```
Comment (fenryka) [inline core/src/test/kotlin/net/corda/core/utilities/KotlinUtilsTest.kt]: no need for full classpath!

File: core/src/test/kotlin/net/corda/core/utilities/KotlinUtilsTest.kt [hunk 2]
```diff
@@ -73,12 +73,12 @@ class KotlinUtilsTest {
...
     private class NonCapturingTransientProperty {
         val transientVal by transient { random63BitValue() }
     }
 
-    @CordaSerializable
+    @net.corda.annotations.serialization.CordaSerializable
```
Comment (fenryka) [inline core/src/test/kotlin/net/corda/core/utilities/KotlinUtilsTest.kt]: no need for full classpath!

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/SerializationHelper.kt [hunk 3]
```diff
@@ -246,8 +246,8 @@ internal fun <T : Any> propertiesForSerializationFromConstructor(
             // with the case we don't know the case of A when the parameter doesn't match a property
             // but has a getter
             val matchingProperty = classProperties[name] ?: classProperties[name.capitalize()] ?:
-                    throw NotSerializableException(
-                            "Constructor parameter - \"$name\" -  doesn't refer to a property of \"$clazz\"")
+            throw NotSerializableException(
```
Comment (fenryka) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/SerializationHelper.kt]: indenting

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/SerializationHelper.kt [hunk 4]
```diff
@@ -268,10 +268,10 @@ internal fun <T : Any> propertiesForSerializationFromConstructor(
                 Pair(PublicPropertyReader(getter), returnType)
             } else {
                 val field = classProperties[name]!!.field ?:
-                        throw NotSerializableException("No property matching constructor parameter named - \"$name\" - " +
-                                "of \"$clazz\". If using Java, check that you have the -parameters option specified " +
-                                "in the Java compiler. Alternately, provide a proxy serializer " +
-                                "(SerializationCustomSerializer) if recompiling isn't an option")
+                throw NotSerializableException("No property matching constructor parameter named - \"$name\" - " +
```
Comment (fenryka) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/SerializationHelper.kt]: indenting

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/SerializationHelper.kt [hunk 5]
```diff
@@ -579,3 +579,6 @@ fun Class<*>.objectInstance() =
                 }
             }
         }
+
+private fun hasSerializationAnnotation(type: Class<*>) =
```
Comment (fenryka) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/SerializationHelper.kt]: second time this functions been implemented, can we refactor to one?

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/carpenter/ClassCarpenter.kt [hunk 6]
```diff
@@ -142,6 +142,7 @@ class ClassCarpenter(cl: ClassLoader = Thread.currentThread().contextClassLoader
 
                 if (schema.flags.cordaSerializable()) {
                     visitAnnotation(Type.getDescriptor(CordaSerializable::class.java), true).visitEnd()
+                    visitAnnotation(Type.getDescriptor(net.corda.annotations.serialization.CordaSerializable::class.java), true).visitEnd()
```
Comment (fenryka) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/carpenter/ClassCarpenter.kt]: why apply both?

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/carpenter/ClassCarpenter.kt [hunk 7]
```diff
@@ -163,6 +164,7 @@ class ClassCarpenter(cl: ClassLoader = Thread.currentThread().contextClassLoader
...
```
Comment (fenryka) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/carpenter/ClassCarpenter.kt]: applying both?

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/carpenter/ClassCarpenter.kt [hunk 8]
```diff
@@ -186,6 +188,7 @@ class ClassCarpenter(cl: ClassLoader = Thread.currentThread().contextClassLoader
...
```
Comment (fenryka) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/carpenter/ClassCarpenter.kt]: applying both

File: node-api/src/test/kotlin/net/corda/nodeapi/internal/AttachmentsClassLoaderTests.kt [hunk 9]
```diff
@@ -207,7 +207,7 @@ class AttachmentsClassLoaderTests {
     }
 
     // top level wrapper
-    @CordaSerializable
+    @net.corda.annotations.serialization.CordaSerializable
```
Comment (fenryka) [inline node-api/src/test/kotlin/net/corda/nodeapi/internal/AttachmentsClassLoaderTests.kt]: fully qualified!

File: node-api/src/test/kotlin/net/corda/nodeapi/internal/serialization/CordaClassResolverTests.kt [hunk 10]
```diff
@@ -23,7 +23,7 @@ import java.sql.Connection
 import java.util.*
 import kotlin.test.*
 
-@CordaSerializable
+@net.corda.annotations.serialization.CordaSerializable
```
Comment (fenryka) [inline node-api/src/test/kotlin/net/corda/nodeapi/internal/serialization/CordaClassResolverTests.kt]: fully qualified!

File: node-api/src/test/kotlin/net/corda/nodeapi/internal/serialization/CordaClassResolverTests.kt [hunk 11]
```diff
@@ -43,7 +43,7 @@ enum class BadFood {
     abstract val value: Int
 }
 
-@CordaSerializable
+@net.corda.annotations.serialization.CordaSerializable
```
Comment (fenryka) [inline node-api/src/test/kotlin/net/corda/nodeapi/internal/serialization/CordaClassResolverTests.kt]: fully qualified!

File: node-api/src/test/kotlin/net/corda/nodeapi/internal/serialization/CordaClassResolverTests.kt [hunk 12]
```diff
@@ -52,7 +52,7 @@ enum class BadSimple {
     Nasty
 }
 
-@CordaSerializable
+@net.corda.annotations.serialization.CordaSerializable
```
Comment (fenryka) [inline node-api/src/test/kotlin/net/corda/nodeapi/internal/serialization/CordaClassResolverTests.kt]: fully qualified!

File: node-api/src/test/kotlin/net/corda/nodeapi/internal/serialization/CordaClassResolverTests.kt [hunk 13]
```diff
@@ -63,7 +63,7 @@ abstract class AbstractClass
 
 interface Interface
 
-@CordaSerializable
+@net.corda.annotations.serialization.CordaSerializable
```
Comment (fenryka) [inline node-api/src/test/kotlin/net/corda/nodeapi/internal/serialization/CordaClassResolverTests.kt]: fully qualified!

File: node-api/src/test/kotlin/net/corda/nodeapi/internal/serialization/CordaClassResolverTests.kt [hunk 14]
```diff
@@ -76,7 +76,7 @@ open class SerializableViaSubInterface : SerializableSubInterface
 
 class SerializableViaSuperSubInterface : SerializableViaSubInterface()
 
-@CordaSerializable
+@net.corda.annotations.serialization.CordaSerializable
```
Comment (fenryka) [inline node-api/src/test/kotlin/net/corda/nodeapi/internal/serialization/CordaClassResolverTests.kt]: fully qualified!

File: node-api/src/test/kotlin/net/corda/nodeapi/internal/serialization/CordaClassResolverTests.kt [hunk 15]
```diff
@@ -85,7 +85,7 @@ class CustomSerializable : KryoSerializable {
     }
 }
 
-@CordaSerializable
+@net.corda.annotations.serialization.CordaSerializable
```
Comment (fenryka) [inline node-api/src/test/kotlin/net/corda/nodeapi/internal/serialization/CordaClassResolverTests.kt]: fully qualified!

File: node-api/src/test/kotlin/net/corda/nodeapi/internal/serialization/CordaClassResolverTests.kt [hunk 16]
```diff
@@ -360,7 +360,7 @@ class CordaClassResolverTests {
         resolver.getRegistration(LinkedHashSet::class.java)
     }
 
-    @CordaSerializable
+    @net.corda.annotations.serialization.CordaSerializable
```
Comment (fenryka) [inline node-api/src/test/kotlin/net/corda/nodeapi/internal/serialization/CordaClassResolverTests.kt]: fully qualified!

File: node-api/src/test/kotlin/net/corda/nodeapi/internal/serialization/amqp/DeserializeAndReturnEnvelopeTests.kt [hunk 17]
```diff
@@ -45,7 +45,7 @@ class DeserializeAndReturnEnvelopeTests {
 
     @Test
     fun unannotatedInterfaceIsNotInSchema() {
-        @CordaSerializable
+        @net.corda.annotations.serialization.CordaSerializable
```
Comment (fenryka) [inline node-api/src/test/kotlin/net/corda/nodeapi/internal/serialization/amqp/DeserializeAndReturnEnvelopeTests.kt]: fully qualified!

File: node/src/integration-test/kotlin/net/corda/node/NodeStatePersistenceTests.kt [hunk 18]
```diff
@@ -32,6 +32,7 @@ import java.lang.management.ManagementFactory
 import kotlin.test.assertEquals
 import kotlin.test.assertNotNull
 
+
```
Comment (fenryka) [inline node/src/integration-test/kotlin/net/corda/node/NodeStatePersistenceTests.kt]: remove newline

**Comment by roastario** [comment]
I hope I've addressed all the formatting / import issues. 

The reason we visit both annotations is that we have to continue supporting the old one for api compatibility. I've just deprecated it and hopefully at some point the future we can remove it after enough time has elapsed.

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/carpenter/ClassCarpenter.kt (hunk 6 above)
Comment (chrisr3) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/carpenter/ClassCarpenter.kt]: Yes, just `import` the new annotation and be done :grinning:.

File: node-api/src/test/kotlin/net/corda/nodeapi/internal/serialization/amqp/DeserializeNeedingCarpentryTests.kt [hunk 19]
```diff
@@ -29,7 +29,7 @@ class DeserializeNeedingCarpentryTests : AmqpCarpenterBase(AllWhitelist) {
 
     private val sf1 = testDefaultFactoryNoEvolution()
 
-    // Deserialize with whitelisting on to check that `CordaSerializable` annotation present.
+    // Deserialize with whitelisting on to check that `net.corda.annotations.serialization.CordaSerializable` annotation present.
```
Comment (chrisr3) [inline node-api/src/test/kotlin/net/corda/nodeapi/internal/serialization/amqp/DeserializeNeedingCarpentryTests.kt]: Why do this? You're changing the `import` so Dokka won't document the wrong one here.

File: node-api/src/test/kotlin/net/corda/nodeapi/internal/serialization/kryo/KryoTests.kt [hunk 20]
```diff
@@ -197,11 +199,11 @@ class KryoTests(private val compression: CordaSerializationEncoding?) {
         assertEquals(-1, readRubbishStream.read())
     }
 
-    @CordaSerializable
```
Comment (chrisr3) [inline node-api/src/test/kotlin/net/corda/nodeapi/internal/serialization/kryo/KryoTests.kt]: Are you still using the old annotation here? If you are then I'd fully-qualify _that_ one and `import` the new one. And if you're not still using the old annotation then just change the `import` to the new one.

File: core/src/main/kotlin/net/corda/core/crypto/CompositeKey.kt [hunk 21]
```diff
@@ -152,7 +152,7 @@ class CompositeKey private constructor(val threshold: Int, children: List<NodeAn
 
         override fun compareTo(other: NodeAndWeight): Int {
             return if (weight == other.weight)
-                // TODO: this might be expensive, consider a faster deterministic compareTo implementation when weights are equal.
+            // TODO: this might be expensive, consider a faster deterministic compareTo implementation when weights are equal.
                 node.encoded.sequence().compareTo(other.node.encoded.sequence())
```
Comment (chrisr3) [inline core/src/main/kotlin/net/corda/core/crypto/CompositeKey.kt]: What's happened here?

**Comment by chrisr3** [comment]
> The reason we visit both annotations is that we have to continue supporting the old one for api compatibility.

Umm, isn't the `ClassCarpenter` _writing_ annotations at that point? I would have thought the Node would only need to be able to _read_ the old annotation to maintain compatibility? (A question for Carpenter gurus...)

**Comment by chrisr3** [comment]
Oh yes, you're also going to need to change `@net.corda.core.serialization.CordaSerializable` to `@net.corda.annotations.serialization.CordaSerializable` everywhere in `.ci/api-current.txt`. _And_ include the new annotation, and mark the old one as `@Deprecated`. You can run the API Stability check locally using:
```bash
$ ./gradlew generateApi
$ .ci/check-api-changes.sh
```

File: node-api/src/test/kotlin/net/corda/nodeapi/internal/serialization/kryo/KryoTests.kt (hunk 20 above)
Comment (fenryka) [reply node-api/src/test/kotlin/net/corda/nodeapi/internal/serialization/kryo/KryoTests.kt]: I'd import the old one as name-depricated just to ram it home as well

**Comment by mikehearn** [comment]
Ouch, that looks like one hell of a merge :( It's been abandoned for over a month.

**Comment by mikehearn** [comment]
Is it faster to just re-attempt this PR again from scratch?

**Comment by roastario** [comment]
Yep, it's much easier than resolving all those conflicts!

**Comment by roastario** [comment]
I'm also experimenting with the Gradle API / Implementation split.

It should be possible to build two jars per project.

This annotation would be the first member of the serialisation API source
set.

**Pull request closed without merging**